lib/rubocop/cop/sorbet/obsolete_strict_memoization.rb

```diff
@@ -0,0 +1,123 @@
+# frozen_string_literal: true
+
+require "rubocop"
+
+module RuboCop
+  module Cop
+    module Sorbet
+      # Checks for the obsolete pattern for initializing instance variables that was required for older Sorbet
+      # versions in `#typed: strict` files.
+      #
+      # It's no longer required, as of Sorbet 0.5.10210
+      # See https://sorbet.org/docs/type-assertions#put-type-assertions-behind-memoization
+      #
+      # @example
+      #
+      #   # bad
+      #   sig { returns(Foo) }
+      #   def foo
+      #     @foo = T.let(@foo, T.nilable(Foo))
+      #     @foo ||= Foo.new
+      #   end
+      #
+      #   # good
+      #   sig { returns(Foo) }
+      #   def foo
+      #     @foo ||= T.let(Foo.new, T.nilable(Foo))
+      #   end
+      #
+      class ObsoleteStrictMemoization < RuboCop::Cop::Base
+        include RuboCop::Cop::MatchRange
+        include RuboCop::Cop::Alignment
+        extend AutoCorrector
+
+        include TargetSorbetVersion
+        minimum_target_sorbet_static_version "0.5.10210"
+
+        MESSAGE = "This two-stage workaround for memoization in `#typed: strict` files is no longer necessary. " \
+          "See https://sorbet.org/docs/type-assertions#put-type-assertions-behind-memoization."
+
+        # @!method legacy_memoization_pattern?(node)
+        def_node_matcher :legacy_memoization_pattern?, <<~PATTERN
+          (begin
+            ...                                           # Match and ignore any other lines that come first.
+            $(ivasgn $_ivar                               # First line: @_ivar = ...
+              (send                                       # T.let(_ivar, T.nilable(_ivar_type))
+                (const nil? :T) :let
+                (ivar _ivar)
+                (send                                     # T.nilable(_ivar_type)
+                  (const nil? :T) :nilable $_ivar_type)))
+            ...
+            $(or-asgn                                     # Second line: @_ivar ||= _initialization_expr
+              (ivasgn _ivar)
+              $_initialization_expr))
+        PATTERN
+
+        def on_begin(node)
+          expression = legacy_memoization_pattern?(node)
+          return unless expression
+
+          first_assignment_node, ivar, ivar_type, second_conditional_assignment_node, initialization_expr = expression
```

Comment:
I tend to express this with a block rather than a guard clause.

```ruby
legacy_memoization_pattern?(node) do |first_assignment_node, ivar, ivar_type, second_conditional_assignment_node, initialization_expr|
  # ...
end
```

Although this is a bunch of captures. Either way works.